src/theory/quantifiers/ceg_t_instantiator.cpp

```diff
@@ -835,7 +842,174 @@ bool EprInstantiator::processEqualTerms( CegInstantiator * ci, SolvedForm& sf, N
   return false;
 }
 
-bool BvInstantiator::processAssertion( CegInstantiator * ci, SolvedForm& sf, Node pv, Node lit, unsigned effort ) {
+Node BvInverter::getSolveVariable( TypeNode tn ) {
+  std::map< TypeNode, Node >::iterator its = d_solve_var.find( tn );
+  if( its==d_solve_var.end() ){
+    Node k = NodeManager::currentNM()->mkSkolem( "s", tn );
+    d_solve_var[tn] = k;
+    return k;
+  }else{
+    return its->second;
+  }
+}
+
+Node BvInverter::getPathToPv( Node lit, Node pv, Node sv, std::vector< unsigned >& path, std::map< TNode, bool >& visited ) {
+  if( visited.find( lit )==visited.end() ){
+    visited[lit] = true;
+    if( lit==pv ){
+      return sv;
+    }else{
+      unsigned rmod = 0; // TODO : randomize?
+      for( unsigned i=0; i<lit.getNumChildren(); i++ ){
+        unsigned ii = ( i + rmod )%lit.getNumChildren();
+        Node litc = getPathToPv( lit[ii], pv, sv, path, visited );
+        if( !litc.isNull() ){
+          // path is outermost term index last
+          path.push_back( ii );
+          std::vector< Node > children;
+          if( lit.getMetaKind() == kind::metakind::PARAMETERIZED ){
+            children.push_back( lit.getOperator() );
+          }
+          for( unsigned j=0; j<lit.getNumChildren(); j++ ){
+            children.push_back( j==ii ? litc : lit[j] );
+          }
+          return NodeManager::currentNM()->mkNode( lit.getKind(), children );
+        }
+      }
+    }
+  }
+  return Node::null();
+}
+
+Node BvInverter::getPathToPv( Node lit, Node pv, Node sv, Node pvs, std::vector< unsigned >& path ) {
+  std::map< TNode, bool > visited;
```

Comment:
`unordered_map` (= hash map) might be a better choice for use cases like that, performance-wise (O(1) vs O(log n) find/insert).

Reply:
If this is significant in practice, I can change my coding style to use unordered_map instead of map.  

However, one generally should not care about optimizing low-level details like this until we actually go to profile CVC4.  In my experience, one can sit down and profile CVC4 for hours and find higher-level algorithmic improvements throughout that lead to actual performance improvements.  I generally have not had changing map -> unordered_map on my radar when profiling in the past, although perhaps I can try this in bottleneck areas.

That being said, if it makes sense to change my coding style to always do this (if the performance gain is obvious), then I am fine using a different kind of map.

Reply:
I in general absolutely agree with @ajreynol regarding code optimization. Usually, the bottleneck is elsewhere and low-level optimizations have not that huge of an impact. However, in particular with (hash) maps (and sets), this can have a huge impact if changed consistently throughout the code. We optimized the hash map / table implementation for Boolector at one point (we use a hand-written implementation there) since we realized that the performance of the implementation was not as good as we thought it is and it had a considerable impact. Very likely that the difference for STL data structures is not as big though, of course. As a matter of good practice, I'd suggest to use unordered_map and unordered_set (if applicable) rather than map. If all agree, I'd also include that in the code style guidelines.

Reply:
I agree.  unordered_map and unordered_set not only perform better, but using them also documents better how you expect the data structure to be used.  I remember some time ago being surprised to discover that map is implemented as a red-black tree under the hood.  This can create significant performance bottlenecks.  And while it's true that premature optimization is dangerous, I don't see any benefit to using map/set over unordered_map/set.  It's easy to change if you later discover you need the ordering.